cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/sql/GcpDatabaseServerUpgradeService.java

```diff
@@ -1,21 +1,96 @@
+import com.sequenceiq.cloudbreak.cloud.gcp.view.GcpDatabaseServerView;
 import com.sequenceiq.cloudbreak.cloud.model.CloudResource;
 import com.sequenceiq.cloudbreak.cloud.model.DatabaseStack;
 import com.sequenceiq.cloudbreak.cloud.notification.PersistenceNotifier;
 import com.sequenceiq.cloudbreak.cloud.template.compute.DatabaseServerUpgradeService;
 import com.sequenceiq.cloudbreak.common.database.TargetMajorVersion;
+import com.sequenceiq.common.api.type.ResourceType;
 
 @Component
 public class GcpDatabaseServerUpgradeService extends GcpDatabaseServerBaseService implements DatabaseServerUpgradeService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcpDatabaseServerUpgradeService.class);
+
+    @Inject
+    private DatabasePollerService databasePollerService;
+
+    @Inject
+    private GcpSQLAdminFactory gcpSQLAdminFactory;
+
+    @Inject
+    private GcpStackUtil gcpStackUtil;
+
     @Override
     public List<CloudResource> upgrade(AuthenticatedContext ac, DatabaseStack stack, PersistenceNotifier resourceNotifier, TargetMajorVersion databaseVersion)
             throws Exception {
+        GcpDatabaseServerView databaseServerView = new GcpDatabaseServerView(stack.getDatabaseServer());
+        String deploymentName = databaseServerView.getDbServerName();
+        SQLAdmin sqlAdmin = gcpSQLAdminFactory.buildSQLAdmin(ac.getCloudCredential(), ac.getCloudCredential().getName());
+        String projectId = gcpStackUtil.getProjectId(ac.getCloudCredential());
+        String availabilityZone = ac.getCloudContext().getLocation().getAvailabilityZone().value();
+        List<CloudResource> buildableResource = new ArrayList<>();
+        buildableResource.add(getGcpDatabase(deploymentName, availabilityZone));
+
+        try {
+            InstancesListResponse list = sqlAdmin.instances().list(projectId).execute();
+            Optional<DatabaseInstance> first = Optional.empty();
+            if (!list.isEmpty()) {
+                first = list.getItems()
+                        .stream()
+                        .filter(e -> e.getName().equals(deploymentName))
+                        .findFirst();
+            }
+            if (!first.isEmpty()) {
+                DatabaseInstance databaseInstance = first.get();
+                databaseInstance.setDatabaseVersion(databaseServerView.getDatabaseType() + "_" + databaseVersion.getVersion());
+                SQLAdmin.Instances.Patch patch = sqlAdmin.instances().patch(projectId, databaseInstance.getName(), databaseInstance);
+                patch.setPrettyPrint(Boolean.TRUE);
+                try {
+                    Operation operation = patch.execute();
+                    verifyOperation(operation, buildableResource);
+                    CloudResource operationAwareCloudResource = createOperationAwareCloudResource(buildableResource.get(0), operation);
+                    databasePollerService.upgradeDatabasePoller(ac, List.of(operationAwareCloudResource));
```

Comment:
How long does a GCP upgrade take? At a first glance it seems the poller waits 30 minutes. Is that enough even if the database is large? Is there a difference between HA and non-HA installations? Does GCP state a timeout for DB upgrade?

Reply:
- How long does a GCP upgrade take?
  With empty database the upgrade took ~15minutes, when we fill the DB with some test data, it took ~20minutes
- At a first glance it seems the poller waits 30 minutes. Is that enough even if the database is large?
  We can increase the timeout if needed, but we don't have information about the size of DB at the customers.
- Is there a difference between HA and non-HA installations?
   In case of GCP we always create the DB in HA mode. So only HA DB upgrade was tested.
- Does GCP state a timeout for DB upgrade?
  I didn't find any information about it. Where should I search for it?

Reply:
Thank you for the answers.
Ok, I think these are generic questions that apply to all cloud platforms. We could maybe add them as a separate jira, as we need to know these parameters.